File: sql/mysql/Oracle/Dart/MySQLParserBase.dart
import 'package:antlr4/antlr4.dart';
import 'dart:io';
import 'dart:core';
import 'dart:convert';
import 'dart:collection';
import 'MySQLLexer.dart';
import 'MySQLParser.dart';
import 'SqlMode.dart';
import 'SqlModes.dart';

/** Base parser class for MySQL parsing. */
abstract class MySQLParserBase extends Parser {

    // To parameterize the parsing process.
    int serverVersion = 0;
    HashSet<SqlMode> sqlModes = HashSet<SqlMode>();
    bool supportMle = true;

    MySQLParserBase(TokenStream input) : super(input)
    {
        this.serverVersion = 80200;
        this.sqlModes = SqlModes.sqlModeFromString("ANSI_QUOTES");
    }

    /// Determines if the given SQL mode is currently active.
    bool isSqlModeActive(SqlMode mode) {
        return sqlModes.contains(mode);
    }

    /// Checks if the parser is in pure identifier mode.
    bool isPureIdentifier() {
        return isSqlModeActive(SqlMode.ansiQuotes);
    }

    /// Checks if the parser is in text string literal mode.
    bool isTextStringLiteral() {
        return !isSqlModeActive(SqlMode.ansiQuotes);
    }

    /// Checks if the parser is handling a stored routine body.
    bool isStoredRoutineBody() {
        return serverVersion >= 80032 && supportMle;
    }

    /// Checks if the parser is handling a SELECT statement with INTO clause.
    bool isSelectStatementWithInto() {
        return serverVersion >= 80024 && serverVersion < 80031;
    }
}

File: sql/mysql/Oracle/Go/MySQLParserBase.go
package parser

import (
	"github.com/antlr4-go/antlr/v4"
)

type MySQLParserBase struct {
    *antlr.BaseParser
    serverVersion int
    sqlModes           map[SqlMode]bool
    supportMle bool
}

var StaticMySQLParserBase MySQLParserBase

func init() {
    StaticMySQLParserBase = MySQLParserBase {
        supportMle:     true,
        serverVersion: 80200,
    }
    StaticMySQLParserBase.sqlModes = sqlModeFromString("ANSI_QUOTES");
}

func NewMySQLParserBase(input antlr.InputStream) *MySQLParserBase {
    r := &MySQLParserBase{
        supportMle:     true,
        serverVersion: 80200,
    }
    r.sqlModes = make(map[SqlMode]bool)
    return r
}

// isSqlModeActive determines if the given SQL mode is currently active in the lexer.
func (m *MySQLParserBase) isSqlModeActive(mode SqlMode) bool {
    return StaticMySQLParserBase.sqlModes[mode]
}

// isPureIdentifier checks if the lexer is in ANSI_QUOTES mode.
func (m *MySQLParserBase) isPureIdentifier() bool {
    return m.isSqlModeActive(AnsiQuotes)
}

// isTextStringLiteral checks if the lexer is not in ANSI_QUOTES mode.
func (m *MySQLParserBase) isTextStringLiteral() bool {
    return !m.isSqlModeActive(AnsiQuotes)
}

// isStoredRoutineBody checks if the server version supports stored routine body.
func (m *MySQLParserBase) isStoredRoutineBody() bool {
    return m.serverVersion >= 80032 && m.supportMle
}

// isSelectStatementWithInto checks if the server version supports SELECT INTO syntax.
func (m *MySQLParserBase) isSelectStatementWithInto() bool {
    return m.serverVersion >= 80024 && m.serverVersion < 80031
}

